Separate NetSim API routes from shared-tables API

islemaster opened the pull request:
Gives NetSim its own API, where before it used our shared-tables API (which is also used for Applab).  For now this API is just a different path into the shared-tables API, but this is an incremental step toward specialized behavior for NetSim, like integrating PubSub functionality and swapping in a new storage layer.

Before, NetSim API URLs looked like this:

```
/v3/shared-tables/<channel-id>/<shard-id>_<table-name>
/v3/shared-tables/JGW2rHUp_UCMW_fQmRf6iQ==/ns_s-netsim-stage-3-puzzle-4_c717e2bd-263a-476c-_n
```

Now they look like this:

```
/v3/netsim/<shard-id>/<table-name>
/v3/netsim/ns_s-netsim-stage-3-puzzle-4_c717e2bd-263a-476c-/n
```

We don't need to provide a channel ID anymore because NetSim always uses an environment-dependent reserved channel, and the server already knows the environment.  Eventually, we won't associate a channel with NetSim at all.

I broke out the shard ID and table name (before they were concatenated to create a table name) for three reasons: 1. I like the way it looks better, 2. our PubSub system will use those values separately, so it's helpful for them to be broken up in the URL, and 3. It declutters the network log in the browser when you see only the one-letter table names:

![screenshot from 2015-07-15 17 59 42](https://cloud.githubusercontent.com/assets/1615761/8713373/bed20ad6-2b1c-11e5-9c9b-60b81b45fd08.png)

islemaster commented:
I don't touch this layer much, so lots of feedback please!
Requesting reviews from:
@philbogle supervising this work
@laurelfan for dashboard experience and possible improvements to test coverage
@Hamms for constant exposure to all things netsim

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,111 @@
+  # Environment-dependent channel ID for NetSim.
+  # Will no longer be needed when we detach from the channels API entirely.
+  def netsim_channel_id
+    rack_env?(:development) ? 'JGW2rHUp_UCMW_fQmRf6iQ==' : 'HQJ8GCCMGP7Yh8MrtDusIA=='
```
Would be better to move this into the config. (That way each environment's config can provide the appropriate value rather than making this code choose the key.)

islemaster replied:
Done in [this commit](https://github.com/code-dot-org/code-dot-org/commit/ef9c484f36459633619573b119207b463b7f75c3).

islemaster commented:
Note: I need to review the NetSimTable tests before finishing this PR, they are passing but I'm not sure they should be.

laurelfan commented on shared/test/test_net_sim_api.rb:
```diff
@@ -0,0 +1,55 @@
+require File.expand_path '../../../deployment', __FILE__
+require File.expand_path '../../middleware/net_sim_api', __FILE__
+
+ENV['RACK_ENV'] = 'test'
```
does this make real api calls? and since it's not in the production env does it affect the production tables?

islemaster replied:
This _does_ make real API calls.  In my local environment this hits my local MySQL-backed channels API, and on test it should hit the test-specific DynamoDB-backed channels API.  If we're uncomfortable with that we should change this and the tables API tests in test_tables.rb.

laurelfan replied:
oh ok, so this line:
`    channel_id = rack_env?(:development) ? 'JGW2rHUp_UCMW_fQmRf6iQ==' : 'HQJ8GCCMGP7Yh8MrtDusIA=='`
does not mean that test and prod are seeing the same channels?

islemaster replied:
If I remember correctly, our `channels_api_secret` ([used here](https://github.com/code-dot-org/code-dot-org/blob/staging/shared/middleware/helpers/storage_id.rb#L18)) is one value for development and a second value for staging/test/production - hence only two public keys here.  This configuration is managed through Chef, though I must admit that whole system is still somewhat opaque to me.

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,111 @@
+  get %r{/v3/netsim/([^/]+)/(\w+)$} do |shard_id, table_name|
+    dont_cache
+    content_type :json
+    TableType.new(netsim_channel_id, storage_id('shared'), channels_table_name(shard_id, table_name)).to_a.to_json
```
To reduce duplicated code, please create a helper method for getting the table given the shard_id and table_name and use it through this class.

islemaster replied:
Done in [this commit](https://github.com/code-dot-org/code-dot-org/commit/05397c53f1db353d0eb2505fa504b9c07c762c7b)

islemaster commented:
JavaScript tests updated.

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,111 @@
+  # Insert a new row.
+  #
+  post %r{/v3/netsim/([^/]+)/(\w+)$} do |shard_id, table_name|
+    unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
```
Is it correct to assume that application/json will always be the first entry in request.content_type?

islemaster replied:
Yes; according to [RFC 2616 Section 3.7](http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.7) the value of the Content-Type field is an "Internet Media Type" given in the form
```
type "/" subtype *( ";" parameter )
```
All we're doing here is removing any optional parameters, which are guaranteed to follow the semicolon.

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,111 @@
+  #
+  post %r{/v3/netsim/([^/]+)/(\w+)$} do |shard_id, table_name|
+    unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
+    unsupported_media_type unless request.content_charset.to_s.downcase == 'utf-8'
```
How about just making this an "and" on the previous line.

islemaster replied:
Done as part of extracting helper method in [this commit](https://github.com/code-dot-org/code-dot-org/commit/8e680dde70557f8e33338475608e10bb1e4b6e52).

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,111 @@
+    unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
+    unsupported_media_type unless request.content_charset.to_s.downcase == 'utf-8'
+
+    value = TableType.new(netsim_channel_id, storage_id('shared'), channels_table_name(shard_id, table_name)).insert(JSON.parse(request.body.read), request.ip)
```
value doesn't seem to be used anywhere;  perhaps just don't assign the return result to a variable?  (Here and elsewhere)

islemaster replied:
`value` is actually used in both methods; in this insert method it shows up on the last line in a string interpolation for the redirect:
```ruby
redirect "/v3/netsim/#{shard_id}/#{table_name}/#{value[:id]}", 301
```

In the update method below it's also used on the last line:
```ruby
value.to_json
```

islemaster replied:
Done in [this commit](https://github.com/code-dot-org/code-dot-org/commit/8e680dde70557f8e33338475608e10bb1e4b6e52).

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,111 @@
+  # Update an existing row.
+  #
+  post %r{/v3/netsim/([^/]+)/(\w+)/(\d+)$} do |shard_id, table_name, id|
+    unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
```
Since this is duplicated code with 78 and 79, please factor out the checks into a helper method so you can just say:

if not is_valid_media_type(request):
   unsupported_media_type

Or you could do it in the single line style used previously.  I prefer the two-line style in this case since unsupported_media_type is an exceptional case.

islemaster replied:
Going with `has_json_utf8_headers(request)` which hides fewer details but may be more accurate - `is_valid_media_type` wouldn't be a true statement for all routes in this file.

Usage now reads:
```ruby
unsupported_media_type unless has_json_utf8_headers(request)
```

We seem to use this one-line early-out pattern in many places, so I'm sticking with it for now.  The helper method certainly makes it read better.

islemaster replied:
Helper method extracted in [this commit](https://github.com/code-dot-org/code-dot-org/commit/8e680dde70557f8e33338475608e10bb1e4b6e52).

philbogle commented on shared/test/test_net_sim_api.rb:
```diff
@@ -0,0 +1,55 @@
+class NetSimApiTest < Minitest::Unit::TestCase
+
+  def create_record(record)
+    @net_sim_api.post "/v3/netsim/#{@shard_id}/#{@table_name}", record.to_json, 'CONTENT_TYPE' => 'application/json;charset=utf-8'
```
Is a helper available for constructing these URLs?  I'm less concerned with the test code but for other production code it would be nice not to scatter knowledge of the format in case we decide to change it.

islemaster replied:
No such helper exists right now, but I can point you to every spot in the code that currently knows how these URLs are constructed.

* On the server
  * net_sim_api.rb which defines the routes.
  * test_net_sim_api.rb which tests the routes.
* On the client
  * [NetSimTable.js](https://github.com/code-dot-org/code-dot-org/pull/3134/files#diff-b4e0cb76b01a931f585c5352404e52afR51) which defines the API base URL
  * [clientApi.js](https://github.com/code-dot-org/code-dot-org/blob/staging/shared/js/clientApi.js) which is still used because, other than the base URL, the NetSim API is still the same as the channels API.  Eventually we may have to break this out too.

Right now I think the only place we could extract some helpers and actually see them used would be for net_sim_api.rb to expose some URL-constructing helpers and use them in the tests - but even then, they would only have one usage since net_sim_api.rb itself uses route regexps.  This seems like a premature optimization.

philbogle commented on shared/test/test_net_sim_api.rb:
```diff
@@ -0,0 +1,55 @@
+
+class NetSimApiTest < Minitest::Unit::TestCase
+
+  def create_record(record)
```
Perhaps make these helpers private and move them to the end with a comment so that it is obvious they are helpers.

islemaster replied:
Made test helpers private in [this commit](https://github.com/code-dot-org/code-dot-org/commit/80d5715d19ef3e07fd0d5937ba23c0812dd472d4).  Went ahead and applied this change to all of the test files in /shared.

philbogle commented on shared/test/test_net_sim_api.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def test_create_read_update_delete
+    # The Tables API does not need to share a cookie jar with the Channels API.
```
I believe this comment is obsolete; do you mean NetSimAPI rather than Tables API?

islemaster replied:
Corrected in the ["make test helpers private" commit](https://github.com/code-dot-org/code-dot-org/commit/80d5715d19ef3e07fd0d5937ba23c0812dd472d4).

philbogle commented:
All done with my initial review pass.

islemaster commented:
@philbogle I believe I've addressed all of your comments from your initial review pass.  Thank you for being so thorough.  Would you like to take another look?

philbogle commented on shared/middleware/net_sim_api.rb:
```diff
@@ -0,0 +1,112 @@
+    unsupported_media_type unless has_json_utf8_headers(request)
+
+    value = get_table(shard_id, table_name).
+        insert(JSON.parse(request.body.read), request.ip)
```
Please catch the JSON::ParserError exception and set a 400 ("Bad Request") status if it is thrown. (By default Sinatra will return the less accurate 500 status code ("server error") which does not convey that the request is invalid and will never succeed.)

islemaster replied:
Tested and implemented: Passing malformed JSON on an INSERT or UPDATE operation results in a 400 response instead of a 500.

philbogle commented on apps/src/netsim/NetSimTable.js:
```diff
@@ -31,17 +31,24 @@ var DEFAULT_POLLING_DELAY_MS = 5000;
+var NetSimTable = module.exports = function (shardID, tableName) {
+  // Require both shardID and tableName to be provided
+  if (!shardID || !tableName) {
+    throw new Error('NetSimTable must be constructed with both arguments. ' +
```
I believe this will also ultimately result in a 500 rather than the more accurate 400 error code; consider adding a custom error handler or catch block to return a 400 status.

islemaster replied:
I'm a little confused; it looks like you're commenting on the JavaScript exception in client code, which does not cause any sort of HTTP request.  Are you saying the server have a custom handler for the sort of malformed URL that would happen if we _didn't_ catch this case?

philbogle replied:
Sorry, that comment makes no sense and no action is needed! (I was trying to find other cases of the net_sim_api.rb comment and got overly enthusiastic.)

philbogle commented:
I had a few couple minor comments about 400 HTTP status codes for invalid requests; after that LGTM!

philbogle commented:
LGTM!
